src/sql/src/plan/func.rs

```diff
@@ -324,27 +324,17 @@ impl<'a> ArgImplementationMatcher<'a> {
         ident: &'a str,
         ecx: &'a ExprContext<'a>,
         impls: &[FuncImpl],
-        args: &[Expr],
+        exprs: Vec<CoercibleScalarExpr>,
+        types: &[Option<ScalarType>],
     ) -> Result<ScalarExpr, failure::Error> {
         // Immediately remove all `impls` we know are invalid.
-        let l = args.len();
+        let l = exprs.len();
         let impls = impls
             .iter()
             .filter(|i| i.params.validate_arg_len(l))
             .collect();
         let mut m = Self { ident, ecx };
 
-        let mut exprs = Vec::new();
-        for arg in args {
-            let expr = query::plan_coercible_expr(ecx, arg)?.0;
-            exprs.push(expr);
-        }
-
-        let types: Vec<_> = exprs
-            .iter()
-            .map(|e| ecx.column_type(e).map(|t| t.scalar_type))
-            .collect();
```

Comment:
Food for thought: is there a way to keep the `types` computation here? I think the reason it needs to live outside is for error message purposes, but maybe it would be possible to generate enough of the nice error message here that you wouldn't need the types in both places? Just a thought! Not sure if feasible.